jsonb-generator/src/main/java/io/avaje/jsonb/generator/ComponentReader.java

```diff
@@ -25,34 +24,35 @@ void read() {
       final TypeElement moduleType = typeElement(fqn);
 
       if (isGeneratedComponent(moduleType)) {
-        var adapters =
-          MetaDataPrism.getInstanceOn(moduleType).value().stream()
-            .map(APContext::asTypeElement)
-            .collect(toList());
-
-        if (adapters.get(0).getModifiers().contains(Modifier.PUBLIC)) {
+        if (hasPublicComponents(moduleType)) {
           componentMetaData.setFullName(fqn);
-          adapters.forEach(t -> readMetaData(moduleType));
-
+          readMetaData(moduleType, componentMetaData);
         } else {
-          // non-public adapters grouped by packageName, does not support generic types (JsonFactory)
+          // non-public adapters grouped by packageName
           var packageName = APContext.elements().getPackageOf(moduleType).getQualifiedName().toString();
           var meta = privateMetaData.computeIfAbsent(packageName, k -> new ComponentMetaData());
-          adapters.stream()
-            .map(TypeElement::getQualifiedName)
-            .map(Object::toString)
-            .forEach(meta::add);
+          readMetaData(moduleType, meta);
         }
       }
     }
   }
 
+  private static boolean hasPublicComponents(TypeElement moduleType) {
+    var firstAdapter =
+      MetaDataPrism.getInstanceOn(moduleType).value().stream()
+        .map(APContext::asTypeElement)
+        .findFirst()
+        .orElseThrow();
+
+    return firstAdapter.getModifiers().contains(Modifier.PUBLIC);
```

Comment:
This matches the existing logic ... but to me isn't quite correct. It would fail if there is a Factory (generic type) but NO normal adapter (for non-generic type).

So I think there should be a followup that addresses that.